// @codingame/monaco-vscode-localization-service-override/vscode/src/vs/workbench/contrib/localization/common/localization.contribution.js
import { Disposable } from 'vscode/vscode/vs/base/common/lifecycle';
import { localize } from 'vscode/vscode/vs/nls';
import { registerAction2 } from 'vscode/vscode/vs/platform/actions/common/actions';
import { SyncDescriptor } from 'vscode/vscode/vs/platform/instantiation/common/descriptors';
import { Registry } from 'vscode/vscode/vs/platform/registry/common/platform';
import { ConfigureDisplayLanguageAction, ClearDisplayLanguageAction } from './localizationsActions.js';
import { Extensions } from 'vscode/vscode/vs/workbench/services/extensionManagement/common/extensionFeatures';
import { ExtensionsRegistry } from 'vscode/vscode/vs/workbench/services/extensions/common/extensionsRegistry';

class BaseLocalizationWorkbenchContribution extends Disposable {
    constructor() {
        super();
        registerAction2(ConfigureDisplayLanguageAction);
        registerAction2(ClearDisplayLanguageAction);
        ExtensionsRegistry.registerExtensionPoint({
            extensionPoint: 'localizations',
            defaultExtensionKind: ['ui', 'workspace'],
            jsonSchema: {
                description: ( localize(7586, "Contributes localizations to the editor")),
                type: 'array',
                default: [],
                items: {
                    type: 'object',
                    required: ['languageId', 'translations'],
                    defaultSnippets: [{ body: { languageId: '', languageName: '', localizedLanguageName: '', translations: [{ id: 'vscode', path: '' }] } }],
                    properties: {
                        languageId: {
                            description: ( localize(7587, 'Id of the language into which the display strings are translated.')),
                            type: 'string'
                        },
                        languageName: {
                            description: ( localize(7588, 'Name of the language in English.')),
                            type: 'string'
                        },
                        localizedLanguageName: {
                            description: ( localize(7589, 'Name of the language in contributed language.')),
                            type: 'string'
                        },
                        translations: {
                            description: ( localize(7590, 'List of translations associated to the language.')),
                            type: 'array',
                            default: [{ id: 'vscode', path: '' }],
                            items: {
                                type: 'object',
                                required: ['id', 'path'],
                                properties: {
                                    id: {
                                        type: 'string',
                                        description: ( localize(
                                            7591,
                                            "Id of VS Code or Extension for which this translation is contributed to. Id of VS Code is always `vscode` and of extension should be in format `publisherId.extensionName`."
                                        )),
                                        pattern: '^((vscode)|([a-z0-9A-Z][a-z0-9A-Z-]*)\\.([a-z0-9A-Z][a-z0-9A-Z-]*))$',
                                        patternErrorMessage: ( localize(
                                            7592,
                                            "Id should be `vscode` or in format `publisherId.extensionName` for translating VS code or an extension respectively."
                                        ))
                                    },
                                    path: {
                                        type: 'string',
                                        description: ( localize(
                                            7593,
                                            "A relative path to a file containing translations for the language."
                                        ))
                                    }
                                },
                                defaultSnippets: [{ body: { id: '', path: '' } }],
                            },
                        }
                    }
                }
            }
        });
    }
}
class LocalizationsDataRenderer extends Disposable {
    constructor() {
        super(...arguments);
        this.type = 'table';
    }
    shouldRender(manifest) {
        return !!manifest.contributes?.localizations;
    }
    render(manifest) {
        const localizations = manifest.contributes?.localizations || [];
        if (!localizations.length) {
            return { data: { headers: [], rows: [] }, dispose: () => { } };
        }
        const headers = [
            ( localize(7594, "Language ID")),
            ( localize(7595, "Language Name")),
            ( localize(7596, "Language Name (Localized)")),
        ];
        const rows = ( (localizations
            .sort((a, b) => a.languageId.localeCompare(b.languageId))
            .map(localization => {
            return [
                localization.languageId,
                localization.languageName ?? '',
                localization.localizedLanguageName ?? ''
            ];
        })));
        return {
            data: {
                headers,
                rows
            },
            dispose: () => { }
        };
    }
}
( (Registry.as(Extensions.ExtensionFeaturesRegistry))).registerExtensionFeature({
    id: 'localizations',
    label: ( localize(7597, "Langauage Packs")),
    access: {
        canToggle: false
    },
    renderer: ( (new SyncDescriptor(LocalizationsDataRenderer))),
});

export { BaseLocalizationWorkbenchContribution };
