// vscode/vscode/src/vs/workbench/contrib/terminal/common/terminalStrings.js
import { localize, localize2 } from '../../../../nls.js';
const terminalStrings = {
    terminal: ( localize(9068, "Terminal")),
    new: ( localize(9069, "New Terminal")),
    doNotShowAgain: ( localize(9070, 'Do Not Show Again')),
    currentSessionCategory: ( localize(9071, 'current session')),
    previousSessionCategory: ( localize(9072, 'previous session')),
    typeTask: ( localize(9073, "Task")),
    typeLocal: ( localize(9074, "Local")),
    actionCategory: ( localize2(9075, "Terminal")),
    focus: ( localize2(9076, "Focus Terminal")),
    focusAndHideAccessibleBuffer: ( localize2(9077, "Focus Terminal and Hide Accessible Buffer")),
    kill: {
        ...( localize2(9078, "Kill Terminal")),
        short: ( localize(9079, "Kill")),
    },
    moveToEditor: ( localize2(9080, "Move Terminal into Editor Area")),
    moveIntoNewWindow: ( localize2(9081, "Move Terminal into New Window")),
    moveToTerminalPanel: ( localize2(9082, "Move Terminal into Panel")),
    changeIcon: ( localize2(9083, "Change Icon...")),
    changeColor: ( localize2(9084, "Change Color...")),
    split: {
        ...( localize2(9085, "Split Terminal")),
        short: ( localize(9086, "Split")),
    },
    unsplit: ( localize2(9087, "Unsplit Terminal")),
    rename: ( localize2(9088, "Rename...")),
    toggleSizeToContentWidth: ( localize2(9089, "Toggle Size to Content Width")),
    focusHover: ( localize2(9090, "Focus Hover")),
    sendSequence: ( localize2(9091, "Send Custom Sequence To Terminal")),
    newWithCwd: ( localize2(9092, "Create New Terminal Starting in a Custom Working Directory")),
    renameWithArgs: ( localize2(9093, "Rename the Currently Active Terminal")),
    scrollToPreviousCommand: ( localize2(9094, "Scroll To Previous Command")),
    scrollToNextCommand: ( localize2(9095, "Scroll To Next Command"))
};
export { terminalStrings };
