// vscode/vscode/src/vs/workbench/contrib/mergeEditor/common/mergeEditor.js
import { localize } from '../../../../nls.js';
import { RawContextKey } from '../../../../platform/contextkey/common/contextkey.js';
const ctxIsMergeEditor = ( (new RawContextKey('isMergeEditor', false, { type: 'boolean', description: ( localize(10824, 'The editor is a merge editor')) })));
const ctxIsMergeResultEditor = ( (new RawContextKey(
 'isMergeResultEditor',
 false,
 { type: 'boolean', description: ( localize(10825, 'The editor is a the result editor of a merge editor.')) }
)));
const ctxMergeEditorLayout = ( (new RawContextKey(
 'mergeEditorLayout',
 'mixed',
 { type: 'string', description: ( localize(10826, 'The layout mode of a merge editor')) }
)));
const ctxMergeEditorShowBase = ( (new RawContextKey(
 'mergeEditorShowBase',
 false,
 { type: 'boolean', description: ( localize(10827, 'If the merge editor shows the base version')) }
)));
const ctxMergeEditorShowBaseAtTop = ( (new RawContextKey(
 'mergeEditorShowBaseAtTop',
 false,
 { type: 'boolean', description: ( localize(10828, 'If base should be shown at the top')) }
)));
const ctxMergeEditorShowNonConflictingChanges = ( (new RawContextKey(
 'mergeEditorShowNonConflictingChanges',
 false,
 { type: 'boolean', description: ( localize(10829, 'If the merge editor shows non-conflicting changes')) }
)));
const ctxMergeBaseUri = ( (new RawContextKey('mergeEditorBaseUri', '', { type: 'string', description: ( localize(10830, 'The uri of the baser of a merge editor')) })));
const ctxMergeResultUri = ( (new RawContextKey('mergeEditorResultUri', '', { type: 'string', description: ( localize(10831, 'The uri of the result of a merge editor')) })));
const StorageCloseWithConflicts = 'mergeEditorCloseWithConflicts';
export { StorageCloseWithConflicts, ctxIsMergeEditor, ctxIsMergeResultEditor, ctxMergeBaseUri, ctxMergeEditorLayout, ctxMergeEditorShowBase, ctxMergeEditorShowBaseAtTop, ctxMergeEditorShowNonConflictingChanges, ctxMergeResultUri };
