// vscode/vscode/src/vs/workbench/contrib/preferences/browser/settingsLayout.js
import { localize } from '../../../../nls.js';
const defaultCommonlyUsedSettings = [
    'files.autoSave',
    'editor.fontSize',
    'editor.fontFamily',
    'editor.tabSize',
    'editor.renderWhitespace',
    'editor.cursorStyle',
    'editor.multiCursorModifier',
    'editor.insertSpaces',
    'editor.wordWrap',
    'files.exclude',
    'files.associations',
    'workbench.editor.enablePreview'
];
function getCommonlyUsedData(toggleData) {
    return {
        id: 'commonlyUsed',
        label: ( localize(10336, "Commonly Used")),
        settings: toggleData?.commonlyUsed ?? defaultCommonlyUsedSettings
    };
}
const tocData = {
    id: 'root',
    label: 'root',
    children: [
        {
            id: 'editor',
            label: ( localize(10337, "Text Editor")),
            settings: ['editor.*'],
            children: [
                {
                    id: 'editor/cursor',
                    label: ( localize(10338, "Cursor")),
                    settings: ['editor.cursor*']
                },
                {
                    id: 'editor/find',
                    label: ( localize(10339, "Find")),
                    settings: ['editor.find.*']
                },
                {
                    id: 'editor/font',
                    label: ( localize(10340, "Font")),
                    settings: ['editor.font*']
                },
                {
                    id: 'editor/format',
                    label: ( localize(10341, "Formatting")),
                    settings: ['editor.format*']
                },
                {
                    id: 'editor/diffEditor',
                    label: ( localize(10342, "Diff Editor")),
                    settings: ['diffEditor.*']
                },
                {
                    id: 'editor/multiDiffEditor',
                    label: ( localize(10343, "Multi-File Diff Editor")),
                    settings: ['multiDiffEditor.*']
                },
                {
                    id: 'editor/minimap',
                    label: ( localize(10344, "Minimap")),
                    settings: ['editor.minimap.*']
                },
                {
                    id: 'editor/suggestions',
                    label: ( localize(10345, "Suggestions")),
                    settings: ['editor.*suggest*']
                },
                {
                    id: 'editor/files',
                    label: ( localize(10346, "Files")),
                    settings: ['files.*']
                }
            ]
        },
        {
            id: 'workbench',
            label: ( localize(10347, "Workbench")),
            settings: ['workbench.*'],
            children: [
                {
                    id: 'workbench/appearance',
                    label: ( localize(10348, "Appearance")),
                    settings: ['workbench.activityBar.*', 'workbench.*color*', 'workbench.fontAliasing', 'workbench.iconTheme', 'workbench.sidebar.location', 'workbench.*.visible', 'workbench.tips.enabled', 'workbench.tree.*', 'workbench.view.*']
                },
                {
                    id: 'workbench/breadcrumbs',
                    label: ( localize(10349, "Breadcrumbs")),
                    settings: ['breadcrumbs.*']
                },
                {
                    id: 'workbench/editor',
                    label: ( localize(10350, "Editor Management")),
                    settings: ['workbench.editor.*']
                },
                {
                    id: 'workbench/settings',
                    label: ( localize(10351, "Settings Editor")),
                    settings: ['workbench.settings.*']
                },
                {
                    id: 'workbench/zenmode',
                    label: ( localize(10352, "Zen Mode")),
                    settings: ['zenmode.*']
                },
                {
                    id: 'workbench/screencastmode',
                    label: ( localize(10353, "Screencast Mode")),
                    settings: ['screencastMode.*']
                }
            ]
        },
        {
            id: 'window',
            label: ( localize(10354, "Window")),
            settings: ['window.*'],
            children: [
                {
                    id: 'window/newWindow',
                    label: ( localize(10355, "New Window")),
                    settings: ['window.*newwindow*']
                }
            ]
        },
        {
            id: 'features',
            label: ( localize(10356, "Features")),
            children: [
                {
                    id: 'features/accessibilitySignals',
                    label: ( localize(10357, 'Accessibility Signals')),
                    settings: ['accessibility.signal*']
                },
                {
                    id: 'features/accessibility',
                    label: ( localize(10358, "Accessibility")),
                    settings: ['accessibility.*']
                },
                {
                    id: 'features/explorer',
                    label: ( localize(10359, "Explorer")),
                    settings: ['explorer.*', 'outline.*']
                },
                {
                    id: 'features/search',
                    label: ( localize(10360, "Search")),
                    settings: ['search.*']
                },
                {
                    id: 'features/debug',
                    label: ( localize(10361, "Debug")),
                    settings: ['debug.*', 'launch']
                },
                {
                    id: 'features/testing',
                    label: ( localize(10362, "Testing")),
                    settings: ['testing.*']
                },
                {
                    id: 'features/scm',
                    label: ( localize(10363, "Source Control")),
                    settings: ['scm.*']
                },
                {
                    id: 'features/extensions',
                    label: ( localize(10364, "Extensions")),
                    settings: ['extensions.*']
                },
                {
                    id: 'features/terminal',
                    label: ( localize(10365, "Terminal")),
                    settings: ['terminal.*']
                },
                {
                    id: 'features/task',
                    label: ( localize(10366, "Task")),
                    settings: ['task.*']
                },
                {
                    id: 'features/problems',
                    label: ( localize(10367, "Problems")),
                    settings: ['problems.*']
                },
                {
                    id: 'features/output',
                    label: ( localize(10368, "Output")),
                    settings: ['output.*']
                },
                {
                    id: 'features/comments',
                    label: ( localize(10369, "Comments")),
                    settings: ['comments.*']
                },
                {
                    id: 'features/remote',
                    label: ( localize(10370, "Remote")),
                    settings: ['remote.*']
                },
                {
                    id: 'features/timeline',
                    label: ( localize(10371, "Timeline")),
                    settings: ['timeline.*']
                },
                {
                    id: 'features/notebook',
                    label: ( localize(10372, 'Notebook')),
                    settings: ['notebook.*', 'interactiveWindow.*']
                },
                {
                    id: 'features/mergeEditor',
                    label: ( localize(10373, 'Merge Editor')),
                    settings: ['mergeEditor.*']
                },
                {
                    id: 'features/chat',
                    label: ( localize(10374, 'Chat')),
                    settings: ['chat.*', 'inlineChat.*']
                }
            ]
        },
        {
            id: 'application',
            label: ( localize(10375, "Application")),
            children: [
                {
                    id: 'application/http',
                    label: ( localize(10376, "Proxy")),
                    settings: ['http.*']
                },
                {
                    id: 'application/keyboard',
                    label: ( localize(10377, "Keyboard")),
                    settings: ['keyboard.*']
                },
                {
                    id: 'application/update',
                    label: ( localize(10378, "Update")),
                    settings: ['update.*']
                },
                {
                    id: 'application/telemetry',
                    label: ( localize(10379, "Telemetry")),
                    settings: ['telemetry.*']
                },
                {
                    id: 'application/settingsSync',
                    label: ( localize(10380, "Settings Sync")),
                    settings: ['settingsSync.*']
                },
                {
                    id: 'application/experimental',
                    label: ( localize(10381, "Experimental")),
                    settings: ['application.experimental.*']
                },
                {
                    id: 'application/other',
                    label: ( localize(10382, "Other")),
                    settings: ['application.*']
                }
            ]
        },
        {
            id: 'security',
            label: ( localize(10383, "Security")),
            settings: ['security.*'],
            children: [
                {
                    id: 'security/workspace',
                    label: ( localize(10384, "Workspace")),
                    settings: ['security.workspace.*']
                }
            ]
        }
    ]
};
const knownAcronyms = ( (new Set()));
[
    'css',
    'html',
    'scss',
    'less',
    'json',
    'js',
    'ts',
    'ie',
    'id',
    'php',
    'scm',
].forEach(str => knownAcronyms.add(str));
const knownTermMappings = ( (new Map()));
knownTermMappings.set('power shell', 'PowerShell');
knownTermMappings.set('powershell', 'PowerShell');
knownTermMappings.set('javascript', 'JavaScript');
knownTermMappings.set('typescript', 'TypeScript');
export { getCommonlyUsedData, knownAcronyms, knownTermMappings, tocData };
