// vscode/vscode/src/vs/workbench/contrib/codeActions/common/codeActionsExtensionPoint.js
import { localize } from '../../../../nls.js';
import { languagesExtPoint } from '../../../services/language/common/languageService.js';
import { Extensions } from '../../../services/extensionManagement/common/extensionFeatures.js';
import { Disposable } from '../../../../base/common/lifecycle.js';
import { SyncDescriptor } from '../../../../platform/instantiation/common/descriptors.js';
import { Registry } from '../../../../platform/registry/common/platform.js';
import { MarkdownString } from '../../../../base/common/htmlContent.js';
var CodeActionExtensionPointFields;
( ((function(CodeActionExtensionPointFields) {
    CodeActionExtensionPointFields["languages"] = "languages";
    CodeActionExtensionPointFields["actions"] = "actions";
    CodeActionExtensionPointFields["kind"] = "kind";
    CodeActionExtensionPointFields["title"] = "title";
    CodeActionExtensionPointFields["description"] = "description";
})(CodeActionExtensionPointFields || (CodeActionExtensionPointFields = {}))));
const codeActionsExtensionPointSchema = ( (Object.freeze({
    type: 'array',
    markdownDescription: ( localize(4552, "Configure which editor to use for a resource.")),
    items: {
        type: 'object',
        required: [CodeActionExtensionPointFields.languages, CodeActionExtensionPointFields.actions],
        properties: {
            [CodeActionExtensionPointFields.languages]: {
                type: 'array',
                description: ( localize(4553, "Language modes that the code actions are enabled for.")),
                items: { type: 'string' }
            },
            [CodeActionExtensionPointFields.actions]: {
                type: 'object',
                required: [CodeActionExtensionPointFields.kind, CodeActionExtensionPointFields.title],
                properties: {
                    [CodeActionExtensionPointFields.kind]: {
                        type: 'string',
                        markdownDescription: ( localize(4554, "`CodeActionKind` of the contributed code action.")),
                    },
                    [CodeActionExtensionPointFields.title]: {
                        type: 'string',
                        description: ( localize(4555, "Label for the code action used in the UI.")),
                    },
                    [CodeActionExtensionPointFields.description]: {
                        type: 'string',
                        description: ( localize(4556, "Description of what the code action does.")),
                    },
                }
            }
        }
    }
})));
const codeActionsExtensionPointDescriptor = {
    extensionPoint: 'codeActions',
    deps: [languagesExtPoint],
    jsonSchema: codeActionsExtensionPointSchema
};
class CodeActionsTableRenderer extends Disposable {
    constructor() {
        super(...arguments);
        this.type = 'table';
    }
    shouldRender(manifest) {
        return !!manifest.contributes?.codeActions;
    }
    render(manifest) {
        const codeActions = manifest.contributes?.codeActions || [];
        if (!codeActions.length) {
            return { data: { headers: [], rows: [] }, dispose: () => { } };
        }
        const flatActions = ( (codeActions.map(
            contribution => ( (contribution.actions.map(action => ({ ...action, languages: contribution.languages }))))
        ))).flat();
        const headers = [
            ( localize(4557, "Title")),
            ( localize(4558, "Kind")),
            ( localize(4559, "Description")),
            ( localize(4560, "Languages"))
        ];
        const rows = ( (flatActions.sort((a, b) => a.title.localeCompare(b.title))
            .map(action => {
            return [
                action.title,
                ( (new MarkdownString())).appendMarkdown(`\`${action.kind}\``),
                action.description ?? '',
                ( (new MarkdownString())).appendMarkdown(`${( (action.languages.map(lang => `\`${lang}\``))).join('&nbsp;')}`),
            ];
        })));
        return {
            data: {
                headers,
                rows
            },
            dispose: () => { }
        };
    }
}
( (Registry.as(Extensions.ExtensionFeaturesRegistry))).registerExtensionFeature({
    id: 'codeActions',
    label: ( localize(4561, "Code Actions")),
    access: {
        canToggle: false,
    },
    renderer: ( (new SyncDescriptor(CodeActionsTableRenderer))),
});
export { codeActionsExtensionPointDescriptor };
